controllers/clustertopology_controller_compute.go

```diff
@@ -69,10 +73,88 @@ type machineDeploymentTopologyState struct {
 	infrastructureMachineTemplate *unstructured.Unstructured   //nolint:structcheck
 }
 
-// Gets the ClusterClass and the referenced templates to be used for a managed Cluster topology.
-func (r *ClusterTopologyReconciler) getClass(ctx context.Context, cluster *clusterv1.Cluster) (*clusterTopologyClass, error) {
-	// TODO: add get class logic; also remove nolint exception from clusterTopologyClass and machineDeploymentTopologyClass
-	return nil, nil
+// getClass gets the ClusterClass and the referenced templates to be used for a managed Cluster topology. It also converts
+// and patches all ObjectReferences in ClusterClass and ControlPlane to the latest apiVersion of the current contract.
+// NOTE: This function assumes that cluster.Spec.Topology.Class is set.
+func (r *ClusterTopologyReconciler) getClass(ctx context.Context, cluster *clusterv1.Cluster) (_ *clusterTopologyClass, reterr error) {
+	class := &clusterTopologyClass{
+		clusterClass:       &clusterv1.ClusterClass{},
+		machineDeployments: map[string]machineDeploymentTopologyClass{},
+	}
+
+	// Get ClusterClass.
+	key := client.ObjectKey{Name: cluster.Spec.Topology.Class, Namespace: cluster.Namespace}
+	if err := r.Client.Get(ctx, key, class.clusterClass); err != nil {
+		return nil, errors.Wrapf(err, "failed to retrieve ClusterClass %q in namespace %q", cluster.Spec.Topology.Class, cluster.Namespace)
+	}
+
+	// We use the patchHelper to patch potential changes to the ObjectReferences in ClusterClass.
+	patchHelper, err := patch.NewHelper(class.clusterClass, r.Client)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		if err := patchHelper.Patch(ctx, class.clusterClass); err != nil {
+			reterr = kerrors.NewAggregate([]error{reterr, errors.Wrapf(err, "failed to patch ClusterClass %q in namespace %q", class.clusterClass.Name, class.clusterClass.Namespace)})
+		}
+	}()
+
+	// Get ClusterClass.spec.infrastructure.
+	class.infrastructureClusterTemplate, err = r.convertReferenceAPIContractAndGet(ctx, class.clusterClass.Spec.Infrastructure.Ref)
```

Comment:
```suggestion
	class.infrastructureClusterTemplate, err = r.getTemplate(ctx, class.clusterClass.Spec.Infrastructure.Ref)
```
And add to the func comment that we are updating the reference if required, so we are consistent with `getClass` and we make it more explicit what the func provides as output (vs surfacing the internal details of how this is achieved).

Reply:
@fabriziopandini I changed it for now to `getTemplate`. We could also name it `getObject` because it's not specific to templates (that's just currently the only situation where we use it). WDYT?

Reply:
I prefer getTemplate, because this is what we are using this method for